### PR TITLE
Loadbalancers

### PR DESCRIPTION
**Pull request by norrland**
CRUD functions created.

TODO:
- [x] managing of blacklists
- [x] managing of certs
- [x] Targets - enable/disable

Examples:
=========

```
// Create the new loadbalancer
newLB, err := client.Loadbalancers.Create(context.Background(),
    glesys.CreateLoadbalancerParams{
        DataCenter: "Falkenberg", Name: "my-awesome-lb", IPv4: "any"})
if err != nil {
    fmt.Println("Error creating: ", err)
}
```

Create backends:
```
// add backend to the new loadbalancer
lbbe, err := client.Loadbalancers.AddBackend(context.Background(), newLB.ID,
    glesys.AddBackendParams{Name: "be-testing123", Mode: "http"})
if err != nil {
    fmt.Println("Error: ", err)
}
```

Adding a target to the backend:
```
// Add a target to the backend
newtarget, err := client.Loadbalancers.AddTarget(context.Background(), newLB.ID,
    glesys.AddTargetParams{Name: "mytarget", Port: 8080, Weight: 5, Backend: "be-testing123", TargetIP: "8.8.8.8"})

if err != nil {
    fmt.Println("Error creating target: ", err)
}
```

Blacklist an ip or prefix in the loadbalancer:
```
lbbl, err := client.Loadbalancers.Addtoblacklist(context.Background(), newLB.ID,
    glesys.BlacklistParams{Prefix: "10.0.0.10/32"})

if err != nil {
    fmt.Println("Error adding blacklist prefix: ", err)
}

### CODE REVIEW

**Comment by wecc** [comment]
@norrland is this still a work in progress?

**Comment by norrland** [comment]
> 
> 
> @norrland is this still a work in progress?

Yes, working out some details on the blacklist functions currently.

**Comment by norrland** [comment]
@wecc Please try it out now.

File: client.go [hunk 1]
```diff
@@ -31,6 +31,7 @@ type Client struct {
 	userAgent  string
 
 	IPs             *IPService
+	Loadbalancers   *LoadbalancerService
```
Comment (wecc) [inline client.go]: Can we rename the `Loadbalancer` instances to `LoadBalancer` throughout the code?

File: loadbalancers.go [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+	Name       string `json:"name"`
+}
+
+// EditLoadbalancerParams is used when editing a loadbalancer
+type EditLoadbalancerParams struct {
+	Name string `json:"name"`
+}
+
+// LBBackend represents a Loadbalancer Backend
+type LBBackend struct {
```
Comment (wecc) [inline loadbalancers.go]: Would love for this to be more explicit: `LoadBalancerBackend`

File: loadbalancers.go [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+}
+
+// AddCertificateParams
+type AddCertificateParams struct {
+	Name        string `json:"certificatename"`
+	Certificate string `json:"certificate"`
+}
+
+// LBFrontend represents a Loadbalancer Frontend
+type LBFrontend struct {
```
Comment (wecc) [inline loadbalancers.go]: `LoadBalancerFrontend`

**Comment by wecc** [comment]
Would you be able to update the CHANGELOG and add some examples in the documentation? (`doc.go` and `doc_test.go`)

File: loadbalancers.go [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+	Targets        []Target `json:"targets"`
+}
+
+// AddBackendParams used when creating backends
+type AddBackendParams struct {
+	ConnectTimeout  int    `json:"connecttimeout,omitempty"`
+	Mode            string `json:"mode,omitempty"`
+	Name            string `json:"name"`
+	ResponseTimeout int    `json:"responsetimeout,omitempty"`
+	Stickysessions  string `json:"stickysessions,omitempty"`
```
Comment (wecc) [inline loadbalancers.go]: `StickySessions` instead of `Stickysessions` (more instances further down)

File: loadbalancers.go [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+	Prefix string `json:"prefix"`
+}
+
+// ToggleTargetParams used when enabling/disabling targets
+type ToggleTargetParams struct {
+	Backend string `json:"backendname"`
+	Name    string `json:"targetname"`
+}
+
+// Create creates a new network
```
Comment (wecc) [inline loadbalancers.go]: typo, should be "creates a new load balancer"

File: loadbalancers.go [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+			Loadbalancer LoadbalancerDetails
+		}
+	}{}
+	return lb.client.post(context, "loadbalancer/addcertificate", &data, struct {
+		AddCertificateParams
+		LoadbalancerID string `json:"loadbalancerid"`
+	}{params, loadbalancerID})
+}
+
+// ListCertificate
```
Comment (wecc) [inline loadbalancers.go]: Missing description

File: loadbalancers.go [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+		}
+	}{}
+	err := lb.client.post(context, "loadbalancer/edittarget", &data, struct {
+		EditTargetParams
+		LoadbalancerID string `json:"loadbalancerid"`
+	}{params, loadbalancerID})
+	return &data.Response.Loadbalancer, err
+}
+
+// Enable a target
```
Comment (wecc) [inline loadbalancers.go]: Should start with `// EnableTarget ...`?

File: loadbalancers.go [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+		}
+	}{}
+	err := lb.client.post(context, "loadbalancer/enabletarget", &data, struct {
+		ToggleTargetParams
+		LoadbalancerID string `json:"loadbalancerid"`
+	}{params, loadbalancerID})
+	return &data.Response.Loadbalancer, err
+}
+
+// Disable a target
```
Comment (wecc) [inline loadbalancers.go]: Should start with `// DisableTarget ...`?

File: loadbalancers.go [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+// RemoveTarget deletes a target
+func (lb *LoadbalancerService) RemoveTarget(context context.Context, loadbalancerID string, params RemoveTargetParams) error {
+	return lb.client.post(context, "loadbalancer/removetarget", nil, struct {
+		RemoveTargetParams
+		LoadbalancerID string `json:"loadbalancerid"`
+	}{params, loadbalancerID})
+}
+
+// Addtoblacklist adds a prefix to loadbalancer blacklist
+func (lb *LoadbalancerService) Addtoblacklist(context context.Context, loadbalancerID string, params BlacklistParams) (*LoadbalancerDetails, error) {
```
Comment (wecc) [inline loadbalancers.go]: `Addtoblacklist` => `AddToBlacklist`

File: loadbalancers.go [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+	}{}
+	err := lb.client.post(context, "loadbalancer/addtoblacklist", &data, struct {
+		BlacklistParams
+		LoadbalancerID string `json:"loadbalancerid"`
+	}{params, loadbalancerID})
+	return &data.Response.Loadbalancer, err
+}
+
+// Removefromblacklist deletes a prefix from loadbalancer blacklist
+func (lb *LoadbalancerService) Removefromblacklist(context context.Context, loadbalancerID string, params BlacklistParams) (*LoadbalancerDetails, error) {
```
Comment (wecc) [inline loadbalancers.go]: `RemoveFromBlacklist`

**Comment by brother** [comment]
Included in release 2.3.0 ac0faaf.

**Pull request closed without merging**